sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```diff
@@ -73,132 +79,136 @@
  * <p>The source also accepts an optional configuration: {@code withQueryFilter()} allows you to
  * define a JSON filter to get subset of files in the database.</p>
  *
- * <p>There is also an optional {@code ParseCallback} that can be specified that can be used to
+ * <p>There is also an optional {@code Parser} that can be specified that can be used to
  * parse the InputStream into objects usable with Beam.  By default, MongoDbGridFSIO will parse
  * into Strings, splitting on line breaks and using the uploadDate of the file as the timestamp.
+ * When using a parser that outputs via outputWithTimestamp, you may also need to specify
+ * the maxSkew option.
  */
 public class MongoDbGridFSIO {
 
   /**
-   * Function for parsing the GridFSDBFile into objects for the PCollection.
+   * Interface for the parser that is used to parse the GridFSDBFile into
+   * the appropriate types.
    * @param <T>
    */
-  public interface ParseCallback<T> extends Serializable {
-    /**
-     * Each value parsed from the file should be output as an
-     * Iterable of Line&lt;T&gt;.  If timestamp is omitted, it will
-     * use the uploadDate of the GridFSDBFile.
-     */
-    public static class Line<T> {
-      final Instant timestamp;
-      final T value;
-
-      public Line(T value, Instant timestamp) {
-        this.value = value;
-        this.timestamp = timestamp;
-      }
-      public Line(T value) {
-        this.value = value;
-        this.timestamp = null;
-      }
-    };
-    public Iterator<Line<T>> parse(GridFSDBFile input) throws IOException;
+  public interface Parser<T> extends Serializable {
+    public void parse(GridFSDBFile input, DoFn<?, T>.ProcessContext result) throws IOException;
```

Comment:
If the parsing can also provide a valid timestamp, how to supply it with only T being passed?     Pass an object that holds the optional timestamp and the value?